src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -28,81 +28,338 @@
 #include <fstream>
 #include <limits>
 #include <sstream>
+#include <tuple>
+#include <stdlib.h>
 #include <unordered_map>
 #include <Corrade/Containers/ArrayView.h>
 #include <Corrade/Utility/String.h>
+#include <Corrade/Utility/Directory.h>
 
 #include "Magnum/Mesh.h"
 #include "Magnum/MeshTools/CombineIndexedArrays.h"
 #include "Magnum/MeshTools/Duplicate.h"
+#include "Magnum/Math/Vector3.h"
 #include "Magnum/Math/Color.h"
 #include "Magnum/Trade/MeshData3D.h"
+#include "Magnum/Trade/MeshObjectData3D.h"
+#include "Magnum/Trade/ImageData.h"
+#include "Magnum/Trade/PhongMaterialData.h"
+
+#include "MagnumPlugins/TgaImporter/TgaImporter.h"
+
+using namespace Corrade::Containers;
 
 namespace Magnum { namespace Trade {
 
-struct ObjImporter::File {
-    std::unordered_map<std::string, UnsignedInt> meshesForName;
-    std::vector<std::string> meshNames;
-    std::vector<std::tuple<std::streampos, std::streampos, UnsignedInt, UnsignedInt, UnsignedInt>> meshes;
-    std::unique_ptr<std::istream> in;
+
+/* Not using PhongMaterialData, since we may want to set color and texture to
+ * later decide which flags we set. We do not know whether we have a texture
+ * or color beforehand. */
+struct ObjMaterial {
+    std::string name;
+    Color3 ambient;
+    Color3 diffuse;
+    Color3 specular;
+    Float specularity;
+    Int ambientTexture = -1;
+    Int diffuseTexture = -1;
+    Int specularTexture = -1;
+};
+
+struct ObjObject;
+struct ObjGroup;
+struct ObjMesh;
+
+struct ObjMeshData {
+    ObjGroup& group;
+    int materialId; /* from 'usemtl' keyword */
```

Comment:
Would be *really* nice if you could use Magnum typedefs for basic types (`Int`, `Float`) .. consistency is king ;)